lib/Reconciler.lua

```diff
@@ -1,542 +1,361 @@
 
-	-- Set properties that were set with fromElement
-	for key, oldValue in pairs(fromElement.props) do
-		seenProps[key] = true
+	DEBUG_warn(("="):rep(80))
+	DEBUG_print("Sync frame:", #tree.tasks, "tasks in queue, marker at", tree.taskIndex)
 
-		local newValue = toElement.props[key]
+	local i = tree.taskIndex
+	while true do
+		local task = tree.tasks[i]
 
-		-- Assume any property that can be set to nil has a default value of nil
-		if newValue == nil then
-			local _, value = getDefaultPropertyValue(rbx.ClassName, key)
+		if task == nil then
+			tree.tasks = {}
+			tree.taskIndex = 1
 
-			-- We don't care if getDefaultPropertyValue fails, because
-			-- _setRbxProp will catch the error below.
-			newValue = value
+			DEBUG_print("Stopping sync frame: ran out of tasks")
+			break
 		end
 
-		-- Roblox does this check for normal values, but we have special
-		-- properties like events that warrant this.
-		if oldValue ~= newValue then
-			Reconciler._setRbxProp(rbx, key, newValue, toElement)
-		end
-	end
-
-	-- Set properties that are new in toElement
-	for key, newValue in pairs(toElement.props) do
-		if not seenProps[key] then
-			seenProps[key] = true
-
-			local oldValue = fromElement.props[key]
+		DEBUG_print("Running task", DEBUG_showTask(task))
+		task(tree)
 
-			if oldValue ~= newValue then
-				Reconciler._setRbxProp(rbx, key, newValue, toElement)
-			end
-		end
+		i = i + 1
 	end
-end
 
---[[
-	Used in _setRbxProp to avoid creating a new closure for every property set.
-]]
-local function set(rbx, key, value)
-	rbx[key] = value
+	tree.tasksRunning = false
 end
 
---[[
-	Sets a property on a Roblox object, following Roact's rules for special
-	case properties.
-
-	This function can throw a couple different errors. In the future, calls to
-	_setRbxProp should be wrapped in a pcall to give better errors to the user.
-
-	For that to be useful, we'll need to attach a 'source' property on every
-	element, created using debug.traceback(), that points to where the element
-	was created.
-]]
-function Reconciler._setRbxProp(rbx, key, value, element)
-	if type(key) == "string" then
-		-- Regular property
+local function mountTree(element, parentRbx, key)
+	assert(Type.is(element, Type.Element))
+	assert(typeof(parentRbx) == "Instance" or typeof(parentRbx) == "nil")
+
+	local tree = {
+		[Type] = Type.Tree,
+		tasks = {},
+		taskIndex = 1,
+		tasksRunning = false,
+		connections = {},
+		mounted = true,
+		rootNode = nil,
+	}
+
+	if ASYNC_SCHEDULER then
+		tree.renderStepId = getGuid()
+		RunService:BindToRenderStep(tree.renderStepId, Enum.RenderPriority.Last.Value, function()
+			processTreeTasksAsync(tree, ASYNC_BUDGET_PER_FRAME)
+		end)
+	end
 
-		local success, err = pcall(set, rbx, key, value)
+	scheduleTask(tree, taskMountNode({
+		element = element,
+		key = key,
+		parentNode = nil,
+		parentRbx = parentRbx,
+		isTreeRoot = true,
+	}))
 
-		if not success then
-			local source = element.source or DEFAULT_SOURCE
+	return tree
```

Comment:
rather than returning a new "tree" which contains internal state, it may not be a terrible idea to make an association table to turn the root node into the internal tree state. This would also fix the "tree" naming issues mentioned below, as all reconciler functions would just take the root node.

Reply:
That's how Roact works today, and I think that was a mistake I'd like to rectify. A handle to a disjoint tree should be explicitly handled differently than a component instance!